src/Runtime/pdf3/internal/generator/browser_session.go

```diff
@@ -0,0 +1,735 @@
+				}
+			}
+		}
+	}
+}
+
+func (w *browserSession) sendCommand(method string, params interface{}) (*cdp.CDPResponse, error) {
+	return w.conn.SendCommand(w.ctx, method, params)
+}
+
+func (w *browserSession) handleRequests() {
+	defer func() {
+		assert.AssertWithMessage(w.ctx.Err() != nil, "Exited worker loop, but process isn't shutting down")
+	}()
+
+	for {
+		select {
+		case <-w.ctx.Done():
+			log.Printf("Worker %d: request processing loop shutting down (context cancellation)\n", w.id)
+			return
+		case req, ok := <-w.queue:
+			if !ok {
+				assert.AssertWithMessage(w.ctx.Err() != nil, "Queue channel closed, but process is not shutting down")
+				return
+			}
+
+			// Reset error counters for this request
+			w.consoleErrors.Store(0)
+			w.browserErrors.Store(0)
+			w.tryUpdateTestModeOutput(&req, "Before", false)
+
+			w.currentUrl = req.request.URL
+			w.handleRequest(&req)
+			w.currentUrl = ""
+
+			if !req.hasResponded() {
+				log.Fatalf("[%d, %s] did not respond to request\n", w.id, w.currentUrl)
+			}
+
+			w.tryUpdateTestModeOutput(&req, "After", true)
+		}
+	}
+}
+
+func (w *browserSession) handleRequest(req *workerRequest) {
+	log.Printf("[%d, %s] processing PDF request\n", w.id, w.currentUrl)
+	start := time.Now()
+
+	defer func() {
+		if r := recover(); r != nil {
+			log.Printf("[%d, %s] recovered from panic: %v\n", w.id, w.currentUrl, r)
+			req.tryRespondError(types.NewPDFError(types.ErrGenerationFail, req.request.URL, fmt.Errorf("%v", r)))
+		}
+
+		duration := time.Since(start)
+		log.Printf("[%d, %s] completed PDF request in %.2f seconds\n", w.id, w.currentUrl, duration.Seconds())
+	}()
+
+	if req.ctx.Err() != nil {
+		req.tryRespondError(types.NewPDFError(types.ErrClientDropped, "", req.ctx.Err()))
+		// When we return here, we never get to the cleanup state in the defer block of `generatePdf` below.
+		// Since we haven't actually used any of the user data the cleanup phase is completely safe to ignore.
+		// So let's manually swap into it cleanup state so that outer loop can swap out of this
+
+		// Capture browser state after cleanup (for test internals mode)
+		w.tryUpdateTestModeOutput(req, "ClientDropped", false)
+		return
+	}
+
+	err := w.generatePdf(req)
+
+	if err != nil {
+		req.tryRespondError(mapCustomError(err))
+	}
+}
+
+func (w *browserSession) generatePdf(req *workerRequest) error {
+	request := req.request
+
+	// Ensure cleanup always runs
+	defer func() {
+		// When we get here we can already accept a request into the queue
+		// Cleanup should run fairly fast (low ms range)
+
+		// Capture browser state before cleanup (for test internals mode)
+		w.tryUpdateTestModeOutput(req, "BeforeCleanup", false)
+
+		// Navigate back to default
+		start := time.Now()
+		_, err := w.sendCommand("Page.navigate", map[string]interface{}{
+			"url": "about:blank",
+		})
+		if err != nil {
+			log.Printf("[%d, %s] failed to navigate out of url: %v\n", w.id, w.currentUrl, err)
+		}
```

Comment:
We are reusing browser pages right? Maybe we should consider resetting navigation history on cleanup 🤔 

https://chromedevtools.github.io/devtools-protocol/tot/Page/#method-resetNavigationHistory